More actions for cesium (3Dtiles geofox), potree, static meshes (f3d), pmtiles, geoparquet

jo-chemla opened the pull request:
Following discussion thread [here](https://github.com/radiantearth/stac-browser/issues/352), adding more actions for the following file types: 

-  OGC 3D-Tiles via Cesium Sandcastle action
-  Potree and COPC via a standalone potree app
-  single-mesh formats like gltf, obj via 3dviewer.net

Will add protomaps pmtiles probably soon. Not completely tested, still a draft for getting directions as to how things look at the moment.

m-mohr commented:
Thanks! Do you have some example items/catalogs/... which I could test this with? I have examples for COPC, but not for the other file formats...

jo-chemla commented:
Hi, thanks for getting back!
 - Just added the simple protomaps pmtiles link/action. 

Our (iconem) [public STAC catalog](http://app.iconem.com/api/stac/collections/project/public/) at the moment only refers to some of the implemented viewers - test url [here](http://localhost:8080/external/app.iconem.com/api/stac/collections/project/public/) for local dev env running), and for other asset types & formats I don't have public data yet: 

- Potree / COPC / ept: [COPC USGS 3DEP ept](http://localhost:8080/external/s3-us-west-2.amazonaws.com/usgs-lidar-stac/ept/ARRA-MO_StoneCo_2010.json?.asset=asset-ept.json) and EPT LoLa Lunar Orbiter [LoLa Lunar Orbiter sample](http://localhost:8080/external/stac.astrogeology.usgs.gov/api/collections/lunar_orbiter_laser_altimeter/items/LolaRDR_-90N-75N_0E15E?.asset=asset-data) USGS Astrogeology STAC LoLa Lunar Orbiter
- pmtiles protomaps: could not find a public reference implementation

I only have private assets for the two below at the moment, will try to open access to one soon and report back. 
- OGC 3D-Tiles via Cesium Sandcastle 
- single-mesh formats like gltf, obj via 3dviewer.net

jo-chemla commented:
Note I also tried these DroneDB (~WebODM wrapper) public STAC catalogs (a bit slow) from [pierotofy brighton-beach](http://localhost:8080/external/hub.dronedb.app/orgs/pierotofy/ds/brighton-beach/stac?.asset=asset-3d-model%2Fodm_textured_model_geo.obj) and [cozy2963 3dhouse](https://radiantearth.github.io/stac-browser/#/external/hub.dronedb.app/orgs/cozy2963/ds/3dhouse/stac?.asset=asset-odm_texturing%2Fodm_textured_model_geo.obj), which include static mesh obj files, but the mesh url seems to be incorrect. 

Also most of the above tests were made with the Allow-Cors plugin switched on to avoid Cors-policy issues.

jo-chemla commented:
@cholmes pointed on linkedin to a open-skysat-data [Planet catalog](https://www.planet.com/data/stac/browser/open-skysat-data/angkor-wat/catalog.json) which stores 3D-tiles tilesets as `external links` (Additional Resources -> 3D Tiles for the Angkor Wat stac-browser page) for the moment rather than `STAC items`, but they might change that soon to make the 3dtiles datasets standard STAC items. This would make it another public STAC of 3D-tiles items while we (iconem) have not published a public collection with 3Dtiles (and in a more distant future static meshes, ). 

As for pmtiles, still could not find any catalog hosting them. The VIDA.place implementation and [deep-dive article](https://medium.com/vida-engineering/building-the-vida-data-catalog-with-stac-and-cloud-native-geospatial-formats-e2240b4977d4) points at an implementation that stores pmtiles and parquet assets, but it seems to be private.

cholmes commented:
> @cholmes pointed on linkedin to a open-skysat-data [Planet catalog](https://www.planet.com/data/stac/browser/open-skysat-data/angkor-wat/catalog.json) which stores 3D-tiles tilesets as external links (Additional Resources -> 3D Tiles for the Angkor Wat stac-browser page) for the moment rather than STAC items, but they might change that soon to make the 3dtiles datasets standard STAC items. This would make it another public STAC of 3D-tiles items while we (iconem) have not published a public collection with 3Dtiles (and in a more distant future static meshes, ).

I wasn't really saying we might change it - more that if the community does coalesce around doing it as assets instead of links then I'd update it. Probably should do something like make a PR in the stac-spec best practices doc at https://github.com/radiantearth/stac-spec/blob/master/best-practices.md#common-media-types-in-stac that says what media type to use for 3d tiles. But does it need to be an asset for STAC Browser? 

> As for pmtiles, still could not find any catalog hosting them. The VIDA.place implementation and [deep-dive article](https://medium.com/vida-engineering/building-the-vida-data-catalog-with-stac-and-cloud-native-geospatial-formats-e2240b4977d4) points at an implementation that stores pmtiles and parquet assets, but it seems to be private

The fiboa STAC's generated by @m-mohr have pmtiles for example https://data.source.coop/fiboa/austria/stac/collection.json and more at https://beta.source.coop/fiboa/ He did it as links too. I think I see more of an argument to do PMTiles as an asset than 3d tiles, mostly because 3d tiles isn't one file - you're linking to the base file that then links to the actual data. I'll raise an issue in stacspec to discuss, we should get to a recommendation for people on what to do.

jo-chemla commented:
Hi there, thanks for the fast reply and sorry for the misunderstanding. Indeed the community has to discuss this, and a PR to the [best-practices.md](https://github.com/radiantearth/stac-spec/blob/master/best-practices.md#common-media-types-in-stac) doc on common-media-types-in-stac might be a great way to introduce this discussion. Should I make an attempt at it? Note that there are not necessarly mime-types for every tileset type described in this stac-browser PR since eg 3dtiles are based on standard json+gltf/glb+textures filetypes. 

From my point-of-view, the reason that 3dtiles datasets could/should be marked as assets is guided by the fact that such datasets are a self-contained geospatial representation of the earth at one point in time. There are even single-file, self-contained COG/COPC 3dtiles packaging equivalent (`.3tz`/`.3dtiles`) described in [this discussion](https://github.com/CesiumGS/3d-tiles/issues/727#issuecomment-2312984585) which would result in a single-file geospatial asset. Even if 3dtiles is a collection of tiles, they don't make sense alone or without the complete hierarchy leading to each tile - the same way for a 2D TMS/XYZ representation of a geotiff, a single tile or subset of tiles would not make sense without the whole tileset. 

We produce 3dtiles for various places on earth, be it monuments or entire cities, and consider them as geospatial assets the same way a pointcloud COPC, a geotiff COG, or a static mesh file, is an asset - each of this dataset types is one byproduct of the 3d-scan processing pipeline. The fact that it is an asset for STAC-Browser, or the qgis plugin, or any STAC reading library means 3dtiles datasets can be imported the same way copc or cog would be imported within an existing project or scene. 

I can see cases where it would make less sense to consider 3dtiles as a single asset, and that is when it is shared as an endpoint to a collection of assets - for example, the Google 3D Cities endpoint. This might make more sense to consider such collections as web-map-links STAC extension, or a link to the endpoint, because the underlying data is a combination of several acquisition and processing. But both points might still hold, this can still be considered a geospatial asset, the same way a mosaick of several input images can still be considered a single asset. 

Plus, thanks a lot for the fiboa pmtiles catalog, will be helpful to test this PR against!

jo-chemla commented:
Following the discussion, I realized how links actions can be created and used within stac-browser and then created them for protomaps and 3d-tiles. I also added your geoparquet.info viewer to geoparquet assets actions. So the following STAC catalogs can be used to test against this PR, and all works great - so will remove the draft, I think it is ready for review: 

Asset actions:
 - Geoparquet: [fiboa STAC](https://data.source.coop/fiboa/austria/stac/collection.json)
 - COPC / Potree / ept: [COPC USGS 3DEP ept](https://s3-us-west-2.amazonaws.com/usgs-lidar-stac/ept/ARRA-MO_StoneCo_2010.json?.asset=asset-ept.json) and EPT LoLa Lunar Orbiter [LoLa Lunar Orbiter sample](https://stac.astrogeology.usgs.gov/api/collections/lunar_orbiter_laser_altimeter/items/LolaRDR_-90N-75N_0E15E?.asset=asset-data) USGS Astrogeology STAC LoLa Lunar Orbiter

Link actions
 - pmtiles protomaps same [fiboa STAC](https://data.source.coop/fiboa/austria/stac/collection.json)
 - OGC 3D-Tiles via Geofox.ai rather than Cesium Sandcastle [Planet catalog](https://www.planet.com/data/stac/open-skysat-data/angkor-wat/catalog.json)

Note pmtiles protomaps were not tested as assets but as links. For OGC3dTiles, we do store them as assets rather than links on our catalog, but it is still private, both asset and links 3dtiles works good. I renamed CesiumSandcastle files and components to OGC3dTiles since the asset action now uses Geofox.ai. And finally regarding 3dviewer.net, there is a bug when parsing mesh url as urlparam, discussions to resolve it are [here](https://github.com/kovacsv/Online3DViewer/issues/432).

m-mohr commented:
Thanks. It may take me some time to review this, but want to let you know I've seen this.

I may extract the geoparquet.info implementation for now as I think it's not really ready for general public usage yet.

jo-chemla commented:
Thanks for getting back, and of course no worries, merging this PR can wait! I've just removed Geoparquet from the assetsActions config - but kept the `./src/actions/assets/Geoparquet.js` file in case it will be useful once geoparquet.info is ready for public use.
 
I'm also considering changing the viewer for 3D static meshes from 3dviewer.net to f3d.app, for which I just implemented a [PR](https://github.com/f3d-app/f3d/pull/1596) to load models passed as url-params. As soon as it is merged I'll update the 3D-viewer asset action.

jo-chemla commented:
Just updated the 3d mesh viewer (from 3dviewer.net to f3d.app/web) since my PR for f3d.app to support url-params has just been merged, because it is more resilient at parsing mesh extension (via either url extension, response contentDisposition header or user-provided extension)!
This PR is ready to go from my side, but no need to rush!

m-mohr commented on src/actions/links/OGC3dTiles.js:
```diff
@@ -0,0 +1,48 @@
+    return this.link.rel === '3d-tiles';
+  }
+
+  get uri_CesiumSandcastle() {
+    let uri = new URI("https://sandcastle.cesium.com/index.html");
+    const tileset_url = this.link.href;
+    const code_payload = {
+      html: `
+        <style> @import url(../templates/bucket.css); </style>
+        <div id="cesiumContainer" class="fullSize"></div> 
+      `, 
+      code: `
+        const viewer = new Cesium.Viewer("cesiumContainer", {
+          terrain: Cesium.Terrain.fromWorldTerrain(),
+        });
+
+        try {
+          const tileset = await Cesium.Cesium3DTileset.fromUrl('${tileset_url}');
+          viewer.scene.primitives.add(tileset);
+          viewer.zoomTo(tileset);
+        } catch (error) {
+          console.log('Error loading tileset');
+        } 
+      `.replaceAll('        ', '')
+    };
+    const code_str = btoa(JSON.stringify(code_payload));
+    uri.addQuery('code', code_str);
+    return uri;
+  }
```
@jo-chemla What is this section doing? It doesn't look like it's used anywhere.

m-mohr replied:
Yeah, I understand that. Sorry, my question was unclear. Where is uri_CesiumSandcastle used? It seems unused...

jo-chemla replied:
See [comment](https://github.com/radiantearth/stac-browser/pull/465#discussion_r1908812396) indeed switched from cesium to geofox.ai for OGC 3D-tiles

jo-chemla replied:
PS Thinking a bit more about it, the Cesium sandcastle can also be seen as more extensible than geofox viewer eg for OGC 3Dtiles pointcloud assets, see this link [https://sandcastle.cesium.com/standalone.html#c=[...]](https://sandcastle.cesium.com/standalone.html#c=3Vl9b9u4Gf8qz2UDzgEcWbIlvzXJ1ibZrkC7FnXaw3Y+4GiJsblKoo+k4viKfPc9pN4tOU3apD3sj6IRyef97Ufa57FUcM3ohgo4gZhu4IxKlkTWB7PWmR/45vuMx4qwmIr5QRc+zWMARYXAlWlOcJl+W1eCRz9zEQbZQuewO49vD5/N43nsG3HSpzFFaalYy3zidkhTTV7zgIaXLKSSKkPFrqDzQyblLWexOgt5EsxWJGDx0mJylqzXXCgadAyrw8NUwQ2LA76xSEiFQjsuV0zCQvCNRFMDTiXEHHVJaWGt+YKvGYNMOc8PtNK3WoOrJPYV4zEIrVQnE1DV31oLFjHFrqm0BI34Ne3s2qK5QcNC9EMSB/QKfRvk4no9uFzR8igoQfyPEhQuSkUUBX4FPBEQsZgBWa9D5hOtn5W7eF24SaKAX+YHM2XBTzSksdQBnB+crRLhr+YHvz7LSUppJ6l59IZE65BebtdUTqscu3rXT4SgsbooD9XO/GL/ao5F5IZFSTTzBaXxbE18eiEEF1NwhpZtTiwpj6gSzDcbMx/jhbvZZkb+XCkaJ8bEKdhdQAdd/J6wazyLUVO8dKEmWhBJ31HJw+SeBHRLz1GJV2y5Uhj4mULLlmpVqrFz4B3mRyLz7dtntRRRaVwv+YfcoZ1sqZI2OxmgKtmOrt0JYZbp5TmrsVdPLWuP1ysc9pzY4dMSG+TRkN52rl2jSiBbGTWPaTYAf3vAWQz4jvB6QrQKrh/ZL7TtXFPgnoRqlbzn7N0c0wy8D7/0ZN5ZiNzGPhS5GnISzFTaF/KupsQ2/QN01cwwP3XbwUajGAnBJxEVxCimyyjk/CMQBa+xvVRajKFO0zgjfMsly5yfdfIzgi1bMhIPzMjQehLUIqUFOOpbzsB1x6OBMx6MxrY7drv5nm2N7ZEzGNv90WgwskejcssbuQNrMHHdyWgy7jvDfraTNt8dpd4Ihu0gz8bK8PuJGl++ZcpfveNhaDQ8p0ssllJDx+lbk4k3GQ5HTt+xB964WyrvWAN3MLE9zx6P+u7QqeiOqg8nnjNxxuPJ2HW8HQ1rUyV1t4UVq7tJ51POJaASg5s1xB0XF6J4ad60xeTiHI2DS4G+v+IiKsb5a4L1fONaL88v/nX58vLf2elskGfZUQ87DM4h72lL1F7gpApAuw5ePZ/BWvBrFuDKYgsrpdZy2uttNhsrZIuQSIuLZU+aYSJ7hQAztLOeBYnEoS3oOsRuFek+LkwLN3+SOIAVkeU0AWo6Hg5HwbFPoCrhFujvCeaw4gV/ndrkGjVd6hRHSJPigIDhmI19vQaU+CujglXNIFX0brIhTBVJbf4bnGduMHnzksfPJX68DDoeZm0eZ41qyvpujlP44eRkZ6IeQpEAujQxQ4hg3Njtr0i8ROwB50z6RAQpWEC4koTKyokEVYmIM/m3LelWATEkCIq5VYT8M9MDvaHH+rP64ft07hoK0sb9zMIQFtTM8gDwwD6RDKOB1fo5kY0RUAVdd5O2D0HnHnaSmoFKJPSzJDvduySrEbaBC8P6Fhs09izomOwv8kXnLMcMDvmy81vqN936tYCMfAp//WRobn/LOO2fGCl2bJ8XBp+DMUnX+VvBryx4S7Gy4HmITaILZxiDaMEIvI8x0QT2rC284wuumC/hFVlgR/FJHGvdkMHzkP4Xa1twuBTc93nITKljd6IMzhjHXA9ZXpqtDQMT2aT0fbpFkW26eBYcc8SowW9Asj8oYIN81HbgDN2+88X9wPn/7gcuEmqzznn8I0YsDPkmNV9CSMQSE0qhjeDCmt3QUFoP7gA4iT0jAcORRDgHYj0xlzhWNcxZo2uMw3ICvHJ5gF5RLNLpjHdJqjZoLuIGLNYFkmKqpApaX9ZSbCvTR1/oMHMRX2FZrJozDbPXJChaVNHQ+vbt6OFYpQKySgBYQCoA17WdketaQ9d1XG/glMgOoN/3bM+xXMcejSeOY3v9yqbrTTy3b1uDkTsau+5gUOwdtgOiO+BeRR8PNRmOBog1PXswGXmjisgj2+qPhkPEc95gPJiMx8PK5tDqjx1UcjzyXNu2x15Dn9vD79jWi4bur6j/8T9U8A465AMJE5qxTxtbvvgPZKzb25oImX6U5w3jtJfsnj8xZYY3qWLcwrR+Jr+Z1K8iz7JnkDMe44xIryDlA0VEo4WuQMxWDh9j7n/kiQK+kFRckwUajqWQhTbftcwTStlgCwkvWFZNldcWbhbO37wGVEcPDKmrX6+9f6m7jsJwIDzR44Cp4s1FcR4uiG56AfcTTYZ1ri5SDi+22O/nB9mZ9GlpV0X9kLPV+mAAZalqN+ec6rxDpX2v5bwprK9Szg+agwSFGyKZLKQv2IJ2imTYzQHIH7wqMyo/YoK7F582wmkyHv2JLetOLs4ulxxtVMfU7Re5Ys+UepA/tO6Z2Cyhm+98FQN29+6YlHsr66tsbpk039Tee86+sm8+jRta3oq+rxdaIU9LL/5603derL6v2Q0Q9iQm73lS+862730UfOrkb30R/HM5I3/PfLArZgiGfSIVoiC8oVzy5TKkLxKleIzT9iLWzoBa1XcNgu2WN9qOSUEaZPo/yNR7GEqaFZ69OdwWP4vdx4iLLSIS9BjkLvv2pjQvA23mHHQPjqXahvQ0D+bfWWR+Y0tE2LGsnqKIRPD+LnuLBKmV5UuZhxbgLzmM+lTmwgKh21Lo6/gUxHJBOm6/C/k/2xqX1DpvAq3sFNz1TWUZ72cBFUci++WmunnbEM3iNeLJigIagzKcVEckZEu8O0QsCELalHqk+HoK/ZrkfAuvcopH9d2maGuFdxFas/6Kx+pooy+ZCOcXPAx2yY97VX8fB+waWHDS8sst+CGREneukjCcsT8QCpwe9/B8gzS7RrxBw0Oy1cdWzumrdNGyrOMefrZTFgi3CP+xRBToKwiIIkcLhLd4iq91wmIgqj8yduFal/kU2jArikv5lHyVLozT0lHHasGDbWVBL4nat14JTl+nwxdSFAYGhoGBI8c93G4S1FdwLc0QhZqhMUI/taB3IxbjF9539N/kBv8euuZDKro2Ow5+VN2Q2bsHFGb+eL9GEjz1oxH6Y8Wz7eooeqO0UIwvfnkPEtlg3nAILojPu7jItDSdkW2lCX8hz9N/Fu8gKWY22PGRI9a/V8BaMOxTBqsVxD9OoIpauCs+X+fSQc2njvnYn5EVPb5B/tegwSN59IV+litB7iN707HrGWrvSdE60n5KV+5eMJ6qgzQQ0JcGqMEIciz+xLHaE6o9F4OnjNneG9Ij1UDTxSnC/1M4OFXlG7o3v3E90Ln4WQM0+F1BPFX4hZg7/0VdlneJFP9ZPo96SxKGVGx7/wM), and still be used as a fullscreen app. Do you have a preference? I'm thinking about switching back from geofox to cesium sandcastle with that fullscreen view.

jo-chemla replied:
Yes good idea, both actions exist in the repo, named after the service used to display the data, but only one will be enabled. Should be done in ~30mins

m-mohr commented on src/actions/assets/F3D.js:
```diff
@@ -0,0 +1,47 @@
+  get show() {
+    return this.component.isBrowserProtocol && (
+      F3D_SUPPORTED_TYPES.includes(this.asset.type)
+      || F3D_SUPPORTED_FILEEXTS.map(
+        f => this.asset.href.toLowerCase().includes(f)
+      ).some(e => e)
```
This would also match for https://www.objects.com/index.html
I think this should use urijs to get the path component and then do an endsWith check.

jo-chemla replied:
I did not know about [objects.com](https://www.objects.com/) - site is under maintenance at the moment, as well as a few previous versions according to wayback machine so I cannot explore it. Is this an open-source 3d-formats-viewer implementation like f3d?

m-mohr replied:
That were just examples for potential false positives because the matching is potentially done on the full URL.

m-mohr commented on src/actions/assets/CopcViewer.js:
```diff
@@ -5,7 +5,11 @@ import i18n from "../../i18n";
-    return this.component.isBrowserProtocol && this.asset.type === 'application/vnd.laszip+copc';
+    return this.component.isBrowserProtocol && (
+      this.asset.type === 'application/vnd.laszip+copc'
+      || this.asset.href.toLowerCase().includes('ept.json')
```
I think we should use urijs to get the path component and then do an endsWith check. That will be more accurate.

jo-chemla replied:
Regarding this comment for Potree, Copc, Protomaps, there are cases where data items are stored on a server with credentials encoded as SAS keys. Maybe relying on the below browser methods would be more like what you have in mind? 

```js
url = new URL ('http://ABC.com/folderpath/ept.json?test=abc')
// URL { href: "http://abc.com/folderpath/ept.json?test=abc", origin: "http://abc.com/", protocol: "http:", username: "", password: "", host: "abc.com", hostname: "abc.com", port: "", pathname: "/folderpath/ept.json", search: "?test=abc" }
check = url.pathname.endsWith('ept.json')
```

m-mohr replied:
Yes, indeed! We just use urijs everywhere in the code, so to keep it consistent you can use:

```
import URI from 'urijs';
const href = 'http://ABC.com/folderpath/ept.json?test=abc';
check = URI(href).filename() == 'ept.json'
```

m-mohr commented on src/actions/assets/OGC3dTiles.js:
```diff
@@ -0,0 +1,55 @@
+    return this.component.isBrowserProtocol && OGC3DTILES_SUPPORTED_TYPES.includes(this.asset.type);
+  }
+
+  get uri_CesiumSandcastle() {
+    let uri = new URI("https://sandcastle.cesium.com/index.html");
+    const tileset_url = this.component.href;
+    const code_payload = {
+      html: `
+        <style> @import url(../templates/bucket.css); </style>
+        <div id="cesiumContainer" class="fullSize"></div> 
+      `, 
+      code: `
+        const viewer = new Cesium.Viewer("cesiumContainer", {
+          terrain: Cesium.Terrain.fromWorldTerrain(),
+        });
+
+        try {
+          const tileset = await Cesium.Cesium3DTileset.fromUrl('${tileset_url}');
+          viewer.scene.primitives.add(tileset);
+          viewer.zoomTo(tileset);
+        } catch (error) {
+          console.log('Error loading tileset');
+        } 
+      `.replaceAll('        ', '')
+    };
+    const code_str = btoa(JSON.stringify(code_payload));
+    uri.addQuery('code', code_str);
+    return uri;
+  }
```
@jo-chemla What is this section doing? It doesn't look like it's used anywhere.

Also, why is this called Cesium Sandcastle but it's using Geofox.ai?

jo-chemla replied:
Sorry I just remembered I switched from Cesium Sandcastle - which is more of a way to test cesium code, although mature and active for years - to the geofox.ai viewer - which looks more like a standalone viewer, focused on loading OGC 3D-tiles tilesets.

m-mohr replied:
Okay, so the code should be removed in both links and assets (for now)?!

jo-chemla replied:
Yes indeed, will do!

m-mohr commented on src/actions/assets/Potree.js:
```diff
@@ -0,0 +1,53 @@
+    return this.component.isBrowserProtocol && (
+      POTREE_SUPPORTED_TYPES.includes(this.asset.type)
+      || POTREE_SUPPORTED_FILEEXTS.map(
+        f => this.asset.href.toLowerCase().includes(f)
```
I think we should use urijs to get the path component and then do an endsWith check. That will be more accurate.

jo-chemla replied:
cf https://github.com/radiantearth/stac-browser/pull/465#discussion_r1908821847

jo-chemla replied:
Just switched to `URI(this.asset.href).filename.endsWith(f)` but below comment might involve a bit more complex check

m-mohr replied:
filename() is a method according to the docs, not a property. Does it work as you posted it? Then there might be some hidden functionality in urijs for property-style access ;-)

jo-chemla replied:
nope i had to use the method indeed

m-mohr commented on src/actions/assets/Protomaps.js:
```diff
@@ -0,0 +1,30 @@
+    // Rather check if .pmtiles substring present in this.asset.href or simply this.component.filename.endsWith('pmtiles')
+    return this.component.isBrowserProtocol && (
+      PROTOMAPS_SUPPORTED_TYPES.includes(this.asset.type)
+      || this.asset.href.toLowerCase().includes('.pmtiles')
```
I think we should use urijs to get the path component and then do an endsWith check. That will be more accurate.

m-mohr replied:
We use urijs everywhere in the code, so to keep it consistent you can use:

```
import URI from 'urijs';
const href = 'http://ABC.com/folderpath/example.pmtiles?v=1';
check = URI(href).suffix() == 'pmtiles'
```

m-mohr reviewed (changes requested):
Some improvements needed for URL parsing and some code seems unused. Otherwise LGTM.

jo-chemla commented:
No worries for the long wait, totally understandable! Just replied to your comments and will quickly edit the PR accordingly!

m-mohr commented on src/actions/assets/Potree.js:
```diff
@@ -0,0 +1,53 @@
+
+// this.component.filename.endsWith or this.asset.href.includes
+const POTREE_SUPPORTED_FILEEXTS = [
+ 'cloud.js', 'metadata.json', 'ept.json'
```
cloud.js and metadata.json are very ambiguous, especially the latter is often used also for non-potree. I think it will lead to a lot of false positives, so not convinced about this approach.

jo-chemla replied:
Are these used for items in a lot of existing STAC catalogs? `cloud.js` it the root file/access-point for a potree v1 dataset, and `metadata.json` is used for potree v2 tilesets. should these be checked differently?

m-mohr replied:
Not sure about cloud.js, but I have at least one examples in mind that use metadata.json for non-potree (CDSE STAC API). And it's pretty expectable as the name is very generic. Can we somehow limit this to when the pointcloud extension is implemented for example?

m-mohr commented:
Thanks, let me know once you are finished :-)

m-mohr commented on src/actions/assets/Potree.js:
```diff
@@ -0,0 +1,53 @@
+    return uri;
+  }
+
+  get uri2() {
+    // Could also use directly Darren Wiens app, which works eg with IGN COPC: 
+    let uri = new URI("https://mpc-copc-viewer.netlify.app");
+    uri.addQuery('c', 'rgba');
+    uri.addQuery('r', this.component.href);
+    // color type can be among rgba, elevation, color, elevation, etc
+    // https://potree.org/potree/examples/copc.html?c=rgba&r=
+    return uri;
+  }
```
That would need to be a separate action, but maybe one action for such a purpose is enough?

As uri2 is currently unused, I'd prefer to remove the code.

jo-chemla commented:
Just did the edits and resolved the corresponding comments. 3 last things: 
 - Regarding potree `metadata.json, cloud.js` filename, how do I check if the pointcloud extension is implemented? 
 - is it possible now to use geoparquet.info or you still consider it as experimental? 
 - Note I'm duplicating a bit of code in `assets` and `links` for Cesium/Geofox.

m-mohr commented:
Thanks for the updates.

>     * Regarding potree `metadata.json, cloud.js` filename, how do I check if the pointcloud extension is implemented?

It seems that's not possible yet as there's no pointer from the asset to the actual "parent" STAC entity.
Will be possible "soon", once I updated to use stac-js. But that's nothing we have available at hand. 
That also means we can't check whether related assets are available.
I'm wondering whether it's really needed to have these links for the metadata documents? Wouldn't you just need those viewers for the main data file? I'm new to point clouds, but it sounds a little weird from an external perspective...
 
>     * is it possible now to use geoparquet.info or you still consider it as experimental?

Actually, no progress since last time. So still experimental.

>     * Note I'm duplicating a bit of code in `assets` and `links` for Cesium/Geofox.

Yeah, I saw that. Not ideal, but fine for now.

m-mohr commented on src/actions/assets/F3D.js:
```diff
@@ -0,0 +1,47 @@
+    // let uri = new URI("https://3dviewer.net/");
+    // uri.addQuery("model", this.component.href); 
+    // uri = uri.toString().replace('?', '#');
+    const extension = ''; // '.obj';
```
That looks suspicious :-) What's the purpose?

jo-chemla replied:
Indeed, a relique from the time where F3D sometime required the extension to be passed this way as url param query.

m-mohr commented on src/actions/assets/Geofox.js:
```diff
@@ -0,0 +1,28 @@
+
+  get uri() {
+    let uri = new URI("https://viewer.geofox.ai/");
+    uri.addQuery('tileset', encodeURI(this.component.href));
```
```suggestion
    uri.addQuery('tileset', this.component.href);
```
urijs encodes the query values automatically, I think.

jo-chemla commented on src/actions/assets/F3D.js:
```diff
@@ -0,0 +1,47 @@
+  'application/fbx',
+];
+// below is usually text/plain
+const F3D_SUPPORTED_FILEEXTS = ['.obj', '.ply', '.fbx', '.glb', '.gltf'];
+export default class F3D extends AssetActionPlugin {
+
+  get show() {
+    return this.component.isBrowserProtocol && (
+      F3D_SUPPORTED_TYPES.includes(this.asset.type)
+      || F3D_SUPPORTED_FILEEXTS.map(
+        ext => URI(this.asset.href).filename().endsWith(ext)
+      ).some(e => e)
+    );
+  }
```
yep indeed

m-mohr commented on src/actions/links/Geofox.js:
```diff
@@ -0,0 +1,21 @@
+
+  get uri() {
+    let uri = new URI("https://viewer.geofox.ai/");
+    uri.addQuery('tileset', encodeURI(this.link.href));
```
```suggestion
    uri.addQuery('tileset', this.link.href);
```

jo-chemla commented:
Thanks for the feedback, understood for geoparquet.info and ~minor duplication in assets/links actions. 

Regarding potree, these metadata documents are actually the pointers to the data itself. Viewers like [potree](http://potree.org/potree/examples/viewer.html), [giro3d](https://giro3d.org/latest/examples/potree_pointcloud.html?dataset=https%3A%2F%2F3d.oslandia.com%2Fpotree%2Fpointclouds%2Flion_takanawa%2Fcloud.js) (soon loadersgl) require either endpoint to the `cloud.js` or `metadata.json` to load it. 

These files are roughly equivalent to the `tileset.json` file of the OGC 3Dtiles spec.

m-mohr commented on assetActions.config.js:
```diff
@@ -1,9 +1,20 @@
-//import Felt from './src/actions/assets/Felt.js';
+// import Geofox from './src/actions/assets/Geofox.js';
+// import Geoparquet from './src/actions/assets/Geoparquet.js';
+import Potree from './src/actions/assets/Potree.js';
```
```suggestion
// import Potree from './src/actions/assets/Potree.js';
```

m-mohr commented on assetActions.config.js:
```diff
@@ -1,9 +1,20 @@
   GeoJsonIo,
-//Felt
+  // Geoparquet, // not ready yet
+  Potree,
```
```suggestion
  // Potree,
```

m-mohr commented on src/actions/assets/Potree.js:
```diff
@@ -0,0 +1,46 @@
+export default class Potree extends AssetActionPlugin {
+
+  get show() {
+    return this.component.isBrowserProtocol && (
```
```suggestion
    // todo: this should check for the pointcloud extension or other indications that this asset is actually point cloud related
    // as the metadata.json matching is rather greedy. Should be possible once stac-js is implemented.
    return this.component.isBrowserProtocol && (
```

jo-chemla commented:
Thanks, just merged your last comments re Potree. 
Ready to merge from my point-of-view, thanks for all the feedback!
